File: server/opensandbox_server/api/lifecycle.py
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
API routes for OpenSandbox Lifecycle API.

This module defines FastAPI routes that map to the OpenAPI specification endpoints.
All business logic is delegated to the service layer that backs each operation.
"""

from typing import List, Optional

from fastapi import APIRouter, Body, Header, HTTPException, Query, Request, status
from fastapi.responses import Response

from opensandbox_server.extensions import validate_extensions
from opensandbox_server.config import get_config
from opensandbox_server.api.schema import (
    CreateSnapshotRequest,
    CreateSandboxRequest,
    CreateSandboxResponse,
    Endpoint,
    ErrorResponse,
    ListSnapshotsRequest,
    ListSnapshotsResponse,
    ListSandboxesRequest,
    ListSandboxesResponse,
    PaginationRequest,
    PatchSandboxMetadataRequest,
    RenewSandboxExpirationRequest,
    RenewSandboxExpirationResponse,
    Sandbox,
    SandboxFilter,
    Snapshot,
    SnapshotFilter,
)
from opensandbox_server.services.constants import (
    OPEN_SANDBOX_INGRESS_HEADER,
    OPEN_SANDBOX_ORIGIN_HEADER,
    SANDBOX_ORIGIN_TEMPLATE,
    SandboxErrorCodes,
)
from opensandbox_server.services.factory import create_sandbox_service
from opensandbox_server.services.snapshot_restore import resolve_sandbox_image_from_request
from opensandbox_server.services.snapshot_service import create_snapshot_service

router = APIRouter(tags=["Sandboxes"])

sandbox_service = create_sandbox_service()
snapshot_service = create_snapshot_service(sandbox_service)
# React to snapshot status changes so rows converge without waiting on reads
snapshot_service.start_background_sync()


# ============================================================================
# Sandbox CRUD Operations
# ============================================================================

@router.post(
    "/sandboxes",
    response_model=CreateSandboxResponse,
    response_model_exclude_none=True,
    status_code=status.HTTP_202_ACCEPTED,
    responses={
        202: {"description": "Sandbox created and provisioned successfully"},
        400: {"model": ErrorResponse, "description": "The request was invalid or malformed"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "Namespace ResourceQuota exhausted — the sandbox was not admitted (KUBERNETES::QUOTA_EXCEEDED)"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        429: {
            "model": ErrorResponse,
            "description": "Pool capacity remained unavailable before the acquisition timeout",
            "headers": {
                "Retry-After": {
                    "description": "Suggested delay in seconds before retrying",
                    "schema": {"type": "integer", "minimum": 1},
                }
            },
        },
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
async def create_sandbox(
    request: CreateSandboxRequest,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> CreateSandboxResponse:
    """
    Create a sandbox from a container image.

    Creates a new sandbox from a container image with optional resource limits,
    environment variables, and metadata. Sandboxes are provisioned directly from
    the specified image without requiring a pre-created template.

    Args:
        request: Sandbox creation request
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        CreateSandboxResponse: Accepted sandbox creation request

    Raises:
        HTTPException: If sandbox creation scheduling fails
    """
    validate_extensions(request.extensions)
    # Resolve snapshotId before backend routing: the owning backend recorded
    # on the snapshot row selects the fsb vs pod backend for restores.
    if not (request.template_id or "").strip():
        request = await resolve_sandbox_image_from_request(request)
    return await sandbox_service.create_sandbox(request)


@router.get(
    "/sandboxes",
    response_model=ListSandboxesResponse,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Paginated collection of sandboxes"},
        400: {"model": ErrorResponse, "description": "The request was invalid or malformed"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def list_sandboxes(
    state: Optional[List[str]] = Query(None, description="Filter by lifecycle state. Pass multiple times for OR logic."),
    metadata: Optional[str] = Query(None, description="Arbitrary metadata key-value pairs for filtering (URL encoded)."),
    page: int = Query(1, ge=1, description="Page number for pagination"),
    page_size: int = Query(20, ge=1, le=200, alias="pageSize", description="Number of items per page"),
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> ListSandboxesResponse:
    """
    List sandboxes with optional filtering and pagination.

    List all sandboxes with optional filtering and pagination using query parameters.
    All filter conditions use AND logic. Multiple `state` parameters use OR logic within states.

    Args:
        state: Filter by lifecycle state.
        metadata: Arbitrary metadata key-value pairs for filtering.
        page: Page number for pagination.
        page_size: Number of items per page.
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        ListSandboxesResponse: Paginated list of sandboxes
    """
    metadata_dict = {}
    if metadata:
        from urllib.parse import parse_qsl
        try:
            # strict_parsing=True rejects malformed segments like "a=1&broken"
            parsed = parse_qsl(metadata, keep_blank_values=True, strict_parsing=True)
            metadata_dict = dict(parsed)
        except Exception as e:
            from fastapi import HTTPException
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={"code": "INVALID_METADATA_FORMAT", "message": f"Invalid metadata format: {str(e)}"}
            )

    request = ListSandboxesRequest(
        filter=SandboxFilter(state=state, metadata=metadata_dict if metadata_dict else None),
        pagination=PaginationRequest(page=page, pageSize=page_size)
    )

    import logging
    logger = logging.getLogger(__name__)
    logger.info(f"ListSandboxes: {request.filter}")

    return sandbox_service.list_sandboxes(request)


@router.get(
    "/sandboxes/{sandbox_id}",
    response_model=Sandbox,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Sandbox current state and metadata"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def get_sandbox(
    sandbox_id: str,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Sandbox:
    """
    Fetch a sandbox by id.

    Returns the complete sandbox information including image specification,
    status, metadata, and timestamps.

    Args:
        sandbox_id: Unique sandbox identifier
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        Sandbox: Complete sandbox information

    Raises:
        HTTPException: If sandbox not found or access denied
    """
    return sandbox_service.get_sandbox(sandbox_id)


@router.patch(
    "/sandboxes/{sandbox_id}/metadata",
    response_model=Sandbox,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Metadata patched successfully. Returns the complete sandbox with updated metadata."},
        400: {"model": ErrorResponse, "description": "The request was invalid or malformed"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def patch_sandbox_metadata(
    sandbox_id: str,
    patch: PatchSandboxMetadataRequest = Body(...),
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Sandbox:
    """
    Patch sandbox metadata via JSON Merge Patch (RFC 7396).
    Non-null adds/replaces, null deletes, absent keeps.
    Read-modify-write without optimistic locking — concurrent PATCH may drop updates.
    """
    return sandbox_service.patch_sandbox_metadata(sandbox_id, patch)


@router.delete(
    "/sandboxes/{sandbox_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    responses={
        204: {"description": "Sandbox successfully deleted"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def delete_sandbox(
    sandbox_id: str,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Response:
    """
    Delete a sandbox.

    Terminates sandbox execution. The sandbox will transition through Stopping state to Terminated.

    Args:
        sandbox_id: Unique sandbox identifier
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        Response: 204 No Content

    Raises:
        HTTPException: If sandbox not found or deletion fails
    """
    sandbox_service.delete_sandbox(sandbox_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


# ============================================================================
# Sandbox Lifecycle Operations
# ============================================================================

@router.post(
    "/sandboxes/{sandbox_id}/pause",
    status_code=status.HTTP_202_ACCEPTED,
    responses={
        202: {"description": "Pause operation accepted"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def pause_sandbox(
    sandbox_id: str,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Response:
    """
    Pause execution while retaining state.

    Pauses a running sandbox while preserving its state.
    Poll GET /sandboxes/{sandboxId} to track state transition through Pausing and eventually Paused.

    Args:
        sandbox_id: Unique sandbox identifier
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        Response: 202 Accepted

    Raises:
        HTTPException: If sandbox not found or cannot be paused
    """
    sandbox_service.pause_sandbox(sandbox_id)
    return Response(status_code=status.HTTP_202_ACCEPTED)


@router.post(
    "/sandboxes/{sandbox_id}/resume",
    status_code=status.HTTP_202_ACCEPTED,
    responses={
        202: {"description": "Resume operation accepted"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def resume_sandbox(
    sandbox_id: str,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Response:
    """
    Resume a paused sandbox.

    Resumes execution of a paused sandbox.
    Poll GET /sandboxes/{sandboxId} to track state transition through Resuming and eventually Running.

    Args:
        sandbox_id: Unique sandbox identifier
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        Response: 202 Accepted

    Raises:
        HTTPException: If sandbox not found or cannot be resumed
    """
    sandbox_service.resume_sandbox(sandbox_id)
    return Response(status_code=status.HTTP_202_ACCEPTED)


@router.post(
    "/sandboxes/{sandbox_id}/renew-expiration",
    response_model=RenewSandboxExpirationResponse,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Sandbox expiration updated successfully"},
        400: {"model": ErrorResponse, "description": "The request was invalid or malformed"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def renew_sandbox_expiration(
    sandbox_id: str,
    request: RenewSandboxExpirationRequest,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> RenewSandboxExpirationResponse:
    """
    Renew sandbox expiration.

    Renews the absolute expiration time of a sandbox.
    The new expiration time must be in the future and after the current expiresAt time.

    Args:
        sandbox_id: Unique sandbox identifier
        request: Renewal request with new expiration time
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        RenewSandboxExpirationResponse: Updated expiration time

    Raises:
        HTTPException: If sandbox not found or renewal fails
    """
    return sandbox_service.renew_expiration(sandbox_id, request)


# ============================================================================
# Snapshot Operations
# ============================================================================

@router.post(
    "/sandboxes/{sandbox_id}/snapshots",
    tags=["Snapshots"],
    response_model=Snapshot,
    response_model_exclude_none=True,
    status_code=status.HTTP_202_ACCEPTED,
    responses={
        202: {"description": "Snapshot creation accepted"},
        400: {"model": ErrorResponse, "description": "The request was invalid or malformed"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The operation conflicts with the current state"},
        501: {"model": ErrorResponse, "description": "Snapshot management is not implemented yet"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def create_snapshot(
    sandbox_id: str,
    response: Response,
    request: Optional[CreateSnapshotRequest] = None,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Snapshot:
    """
    Create a persistent point-in-time snapshot from a sandbox.
    """
    create_request = request or CreateSnapshotRequest()
    snapshot = snapshot_service.create_snapshot(sandbox_id, create_request)
    response.headers["Location"] = f"/v1/snapshots/{snapshot.id}"
    return snapshot


@router.get(
    "/snapshots",
    tags=["Snapshots"],
    response_model=ListSnapshotsResponse,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Paginated collection of snapshots"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        501: {"model": ErrorResponse, "description": "Snapshot management is not implemented yet"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def list_snapshots(
    sandbox_id: Optional[str] = Query(None, alias="sandboxId", description="Filter snapshots by source sandbox identifier"),
    name: Optional[str] = Query(None, description="Filter snapshots by exact snapshot name"),
    state: Optional[List[str]] = Query(None, description="Filter by snapshot lifecycle state. Pass multiple times for OR logic."),
    page: int = Query(1, ge=1, description="Page number for pagination"),
    page_size: int = Query(20, ge=1, le=200, alias="pageSize", description="Number of items per page"),
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> ListSnapshotsResponse:
    """
    List snapshots with optional filtering and pagination.
    """
    request = ListSnapshotsRequest(
        filter=SnapshotFilter(sandboxId=sandbox_id, name=name, state=state),
        pagination=PaginationRequest(page=page, pageSize=page_size),
    )
    return snapshot_service.list_snapshots(request)


@router.get(
    "/snapshots/{snapshot_id}",
    tags=["Snapshots"],
    response_model=Snapshot,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Snapshot current state and metadata"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        501: {"model": ErrorResponse, "description": "Snapshot management is not implemented yet"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def get_snapshot(
    snapshot_id: str,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Snapshot:
    """
    Fetch a snapshot by id.
    """
    return snapshot_service.get_snapshot(snapshot_id)


@router.delete(
    "/snapshots/{snapshot_id}",
    tags=["Snapshots"],
    status_code=status.HTTP_204_NO_CONTENT,
    responses={
        204: {"description": "Snapshot successfully deleted"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        409: {"model": ErrorResponse, "description": "The snapshot is not in a deletable state or is still in use"},
        501: {"model": ErrorResponse, "description": "Snapshot management is not implemented yet"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def delete_snapshot(
    snapshot_id: str,
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
) -> Response:
    """
    Delete a snapshot by id.
    """
    snapshot_service.delete_snapshot(snapshot_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


# ============================================================================
# Sandbox Endpoints
# ============================================================================

@router.get(
    "/sandboxes/{sandbox_id}/endpoints/{port}",
    response_model=Endpoint,
    response_model_exclude_none=True,
    responses={
        200: {"description": "Endpoint retrieved successfully"},
        400: {"model": ErrorResponse, "description": "The request was invalid or malformed"},
        401: {"model": ErrorResponse, "description": "Authentication credentials are missing or invalid"},
        403: {"model": ErrorResponse, "description": "The authenticated user lacks permission for this operation"},
        404: {"model": ErrorResponse, "description": "The requested resource does not exist"},
        500: {"model": ErrorResponse, "description": "An unexpected server error occurred"},
    },
)
def get_sandbox_endpoint(
    request: Request,
    sandbox_id: str,
    port: int,
    use_server_proxy: bool = Query(False, description="Whether to return a server-proxied URL"),
    expires: Optional[int] = Query(None, description="Request a signed route token with this Unix epoch second expiration. Requires ingress gateway with secure_access configured."),
    x_request_id: Optional[str] = Header(None, alias="X-Request-ID", description="Unique request identifier for tracing"),
    response: Response = None,  # type: ignore[assignment]
) -> Endpoint:
    """
    Get sandbox access endpoint.

    Returns the public access endpoint URL for accessing a service running on a specific port
    within the sandbox. The service must be listening on the specified port inside the sandbox
    for the endpoint to be available.

    When the ``expires`` query parameter is provided, the endpoint is wrapped in a
    cryptographically signed route token (OSEP-0011) instead of returning a plain URL.
    This requires the ingress gateway to be configured with secure_access signing keys.

    Args:
        request: FastAPI request object
        sandbox_id: Unique sandbox identifier
        port: Port number where the service is listening inside the sandbox (1-65535)
        use_server_proxy: Whether to return a server-proxied URL
        expires: Unix epoch seconds for signed route token expiration. Must be a
            non-negative uint64 value. When omitted or invalid, a plain (unsigned)
            endpoint is returned.
        x_request_id: Unique request identifier for tracing (optional; server generates if omitted).

    Returns:
        Endpoint: Public endpoint URL

    Raises:
        HTTPException: If sandbox not found, endpoint not available, or signed
            routes are not supported by the runtime/configuration (400).
    """
    if use_server_proxy and expires is not None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "code": SandboxErrorCodes.INVALID_PARAMETER,
                "message": (
                    "use_server_proxy cannot be combined with expires; "
                    "signed endpoints require the gateway route."
                ),
            },
        )

    endpoint = sandbox_service.get_endpoint(sandbox_id, port, expires=expires)

    if use_server_proxy:
        # Prefer configured external address when available (it is the full
        # external base and may already carry a path prefix).
        eip = (get_config().server.eip or "").strip().rstrip("/")
        if eip:
            base_url = eip
        else:
            # Reflect how the client reached this API: same host AND same mount
            # prefix (the API is served both bare and under /v1). Dropping the
            # prefix pointed proxied traffic at whatever backend owns the root
            # path on shared hosts (breaking HTTP with redirects and WebSocket
            # handshakes outright).
            mount_prefix = request.url.path.split("/sandboxes/", 1)[0]
            base_url = str(request.base_url).rstrip("/") + mount_prefix
        base_url = base_url.replace("https://", "").replace("http://", "")
        endpoint.endpoint = f"{base_url}/sandboxes/{sandbox_id}/proxy/{port}"
        if endpoint.headers:
            endpoint.headers = {
                key: value
                for key, value in endpoint.headers.items()
                if key.lower() != OPEN_SANDBOX_INGRESS_HEADER.lower()
            } or None

    # Tell clients the origin of this sandbox. fsb sandboxes (id prefix,
    # mirroring CompositeSandboxService._backend routing) run on golden-image
    # templates and have no sandbox-side egress sidecar; the value space may
    # grow (image/snapshot) as server-side origin tracking matures. Clients
    # treat a missing/unknown value as "not template".
    if response is not None and sandbox_id.startswith("fsb-"):
        response.headers[OPEN_SANDBOX_ORIGIN_HEADER] = SANDBOX_ORIGIN_TEMPLATE

    return endpoint

File: sdks/sandbox/go/lifecycle.go
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package opensandbox

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// LifecycleClient provides methods for the OpenSandbox Lifecycle API.
type LifecycleClient struct {
	*Client
	cache *EndpointCache
}

// NewLifecycleClient creates a new LifecycleClient.
// baseURL should include the version prefix (e.g. "http://localhost:8080/v1").
func NewLifecycleClient(baseURL, apiKey string, opts ...Option) *LifecycleClient {
	return &LifecycleClient{
		Client: NewClient(baseURL, apiKey, "OPEN-SANDBOX-API-KEY", opts...),
		cache:  NewEndpointCache(0, 0),
	}
}

// NewLifecycleClientWithCache creates a LifecycleClient with custom cache settings.
func NewLifecycleClientWithCache(baseURL, apiKey string, cache *EndpointCache, opts ...Option) *LifecycleClient {
	return &LifecycleClient{
		Client: NewClient(baseURL, apiKey, "OPEN-SANDBOX-API-KEY", opts...),
		cache:  cache,
	}
}

// ListOptions configures filtering and pagination for ListSandboxes.
type ListOptions struct {
	// States filters by lifecycle state. Multiple values use OR logic.
	States []SandboxState
	// Metadata filters by key-value metadata (AND logic).
	Metadata map[string]string
	// Page number (1-based). Defaults to 1.
	Page int
	// PageSize is the number of items per page. Defaults to 20.
	PageSize int
}

// ListSandboxes returns a paginated list of sandboxes with optional filtering.
func (c *LifecycleClient) ListSandboxes(ctx context.Context, opts ListOptions) (*ListSandboxesResponse, error) {
	params := url.Values{}
	for _, s := range opts.States {
		params.Add("state", string(s))
	}
	if len(opts.Metadata) > 0 {
		metaVals := url.Values{}
		for k, v := range opts.Metadata {
			metaVals.Set(k, v)
		}
		params.Set("metadata", metaVals.Encode())
	}
	if opts.Page > 0 {
		params.Set("page", strconv.Itoa(opts.Page))
	}
	if opts.PageSize > 0 {
		params.Set("pageSize", strconv.Itoa(opts.PageSize))
	}

	path := "/sandboxes"
	if encoded := params.Encode(); encoded != "" {
		path += "?" + encoded
	}

	var resp ListSandboxesResponse
	if err := c.doRequest(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// CreateSandbox creates a new sandbox from a container image.
func (c *LifecycleClient) CreateSandbox(ctx context.Context, req CreateSandboxRequest) (*SandboxInfo, error) {
	var resp SandboxInfo
	if err := c.doRequest(ctx, "POST", "/sandboxes", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// CreateTemplate declares a new fsb template (golden-image build).
// The build runs asynchronously: the response starts at
// TemplatePhasePending; poll GetTemplate until the phase is Succeeded (or
// Failed). Only Succeeded templates can back template-based sandbox
// creation.
func (c *LifecycleClient) CreateTemplate(ctx context.Context, req CreateTemplateRequest) (*TemplateInfo, error) {
	var resp TemplateInfo
	if err := c.doRequest(ctx, "POST", "/templates", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetTemplate retrieves one template with its latest build status by ID.
func (c *LifecycleClient) GetTemplate(ctx context.Context, templateID string) (*TemplateInfo, error) {
	var resp TemplateInfo
	if err := c.doRequest(ctx, "GET", "/templates/"+url.PathEscape(templateID), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ListTemplates returns a paginated list of templates with optional filtering.
func (c *LifecycleClient) ListTemplates(ctx context.Context, opts ListTemplatesOptions) (*ListTemplatesResponse, error) {
	params := url.Values{}
	if len(opts.Metadata) > 0 {
		metaVals := url.Values{}
		for k, v := range opts.Metadata {
			metaVals.Set(k, v)
		}
		params.Set("metadata", metaVals.Encode())
	}
	if opts.Page > 0 {
		params.Set("page", strconv.Itoa(opts.Page))
	}
	if opts.PageSize > 0 {
		params.Set("pageSize", strconv.Itoa(opts.PageSize))
	}

	path := "/templates"
	if encoded := params.Encode(); encoded != "" {
		path += "?" + encoded
	}

	var resp ListTemplatesResponse
	if err := c.doRequest(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// DeleteTemplate deletes a template by ID. Sandboxes already created from the
// template are unaffected.
func (c *LifecycleClient) DeleteTemplate(ctx context.Context, templateID string) error {
	return c.doRequest(ctx, "DELETE", "/templates/"+url.PathEscape(templateID), nil, nil)
}

// GetNetworkPolicy returns the sandbox's egress policy intent from the
// lifecycle control plane. This is the template-backed path for egress
// policy: fsb sandboxes have no sandbox-side egress sidecar.
func (c *LifecycleClient) GetNetworkPolicy(ctx context.Context, sandboxID string) (*PolicyStatusResponse, error) {
	var resp PolicyStatusResponse
	if err := c.doRequest(ctx, "GET", "/sandboxes/"+url.PathEscape(sandboxID)+"/networkpolicy", nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// PatchNetworkPolicy merges egress rules into the sandbox's policy using the
// same semantics as the sandbox-side egress service: incoming rules take
// priority over existing rules with the same target and replace them in
// place; within one patch payload, the first rule for a target wins;
// existing rules for other targets and the current defaultAction are
// preserved.
func (c *LifecycleClient) PatchNetworkPolicy(ctx context.Context, sandboxID string, rules []NetworkRule) (*PolicyStatusResponse, error) {
	var resp PolicyStatusResponse
	if err := c.doRequest(ctx, "PATCH", "/sandboxes/"+url.PathEscape(sandboxID)+"/networkpolicy", rules, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// DeleteNetworkPolicyRules removes the sandbox's egress rules whose target
// matches one of the given entries. Matching is by exact target string;
// unknown targets are silently ignored (idempotent). The current
// defaultAction is preserved.
func (c *LifecycleClient) DeleteNetworkPolicyRules(ctx context.Context, sandboxID string, targets []string) (*PolicyStatusResponse, error) {
	var resp PolicyStatusResponse
	if err := c.doRequest(ctx, "DELETE", "/sandboxes/"+url.PathEscape(sandboxID)+"/networkpolicy", targets, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// ListSnapshots returns a paginated list of snapshots with optional filtering.
func (c *LifecycleClient) ListSnapshots(ctx context.Context, opts ListSnapshotsOptions) (*ListSnapshotsResponse, error) {
	params := url.Values{}
	if opts.SandboxID != "" {
		params.Set("sandboxId", opts.SandboxID)
	}
	if opts.Name != "" {
		params.Set("name", opts.Name)
	}
	for _, s := range opts.States {
		params.Add("state", string(s))
	}
	if opts.Page > 0 {
		params.Set("page", strconv.Itoa(opts.Page))
	}
	if opts.PageSize > 0 {
		params.Set("pageSize", strconv.Itoa(opts.PageSize))
	}

	path := "/snapshots"
	if encoded := params.Encode(); encoded != "" {
		path += "?" + encoded
	}

	var resp ListSnapshotsResponse
	if err := c.doRequest(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// CreateSnapshot creates a snapshot from the given sandbox.
func (c *LifecycleClient) CreateSnapshot(ctx context.Context, sandboxID string, req CreateSnapshotRequest) (*SnapshotInfo, error) {
	var resp SnapshotInfo
	if err := c.doRequest(ctx, "POST", "/sandboxes/"+url.PathEscape(sandboxID)+"/snapshots", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetSnapshot retrieves snapshot information by ID.
func (c *LifecycleClient) GetSnapshot(ctx context.Context, id string) (*SnapshotInfo, error) {
	var resp SnapshotInfo
	if err := c.doRequest(ctx, "GET", "/snapshots/"+url.PathEscape(id), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// DeleteSnapshot deletes a snapshot by ID.
func (c *LifecycleClient) DeleteSnapshot(ctx context.Context, id string) error {
	return c.doRequest(ctx, "DELETE", "/snapshots/"+url.PathEscape(id), nil, nil)
}

// GetSandbox retrieves the complete sandbox information by ID.
func (c *LifecycleClient) GetSandbox(ctx context.Context, id string) (*SandboxInfo, error) {
	var resp SandboxInfo
	if err := c.doRequest(ctx, "GET", "/sandboxes/"+url.PathEscape(id), nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// PatchSandboxMetadata patches metadata for a sandbox. Non-nil values add or
// replace keys. Nil values delete keys.
func (c *LifecycleClient) PatchSandboxMetadata(ctx context.Context, id string, patch MetadataPatch) (*SandboxInfo, error) {
	var resp SandboxInfo
	if err := c.doRequest(ctx, "PATCH", "/sandboxes/"+url.PathEscape(id)+"/metadata", patch, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// DeleteSandbox deletes a sandbox, scheduling it for termination.
func (c *LifecycleClient) DeleteSandbox(ctx context.Context, id string) error {
	return c.doRequest(ctx, "DELETE", "/sandboxes/"+url.PathEscape(id), nil, nil)
}

// PauseSandbox pauses a running sandbox while preserving its state.
func (c *LifecycleClient) PauseSandbox(ctx context.Context, id string) error {
	return c.doRequest(ctx, "POST", "/sandboxes/"+url.PathEscape(id)+"/pause", nil, nil)
}

// ResumeSandbox resumes a paused sandbox.
func (c *LifecycleClient) ResumeSandbox(ctx context.Context, id string) error {
	return c.doRequest(ctx, "POST", "/sandboxes/"+url.PathEscape(id)+"/resume", nil, nil)
}

// RenewExpiration renews the sandbox's absolute expiration time.
// The caller is responsible for computing the desired expiresAt time.
func (c *LifecycleClient) RenewExpiration(ctx context.Context, id string, expiresAt time.Time) (*RenewExpirationResponse, error) {
	req := RenewExpirationRequest{
		ExpiresAt: expiresAt.UTC(),
	}
	var resp RenewExpirationResponse
	if err := c.doRequest(ctx, "POST", "/sandboxes/"+url.PathEscape(id)+"/renew-expiration", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetEndpoint retrieves the public access endpoint for a service running
// on the specified port inside the sandbox. Results are cached with LRU+TTL
// and deduplicated via singleflight. If useServerProxy is non-nil,
// the server proxy query parameter is included.
func (c *LifecycleClient) GetEndpoint(ctx context.Context, sandboxID string, port int, useServerProxy *bool) (*Endpoint, error) {
	if c.cache == nil {
		return c.getEndpointFromServer(ctx, sandboxID, port, useServerProxy)
	}

	key := endpointCacheKey{
		sandboxID:      sandboxID,
		port:           port,
		useServerProxy: useServerProxy != nil && *useServerProxy,
	}

	// The shared fetch uses a background context so one caller's deadline
	// doesn't cancel the request for all waiters. Each caller's ctx is
	// respected via DoChan select in GetOrFetch.
	return c.cache.GetOrFetch(ctx, key, func() (*Endpoint, error) {
		return c.getEndpointFromServer(context.Background(), sandboxID, port, useServerProxy)
	})
}

func (c *LifecycleClient) getEndpointFromServer(ctx context.Context, sandboxID string, port int, useServerProxy *bool) (*Endpoint, error) {
	path := fmt.Sprintf("/sandboxes/%s/endpoints/%d", url.PathEscape(sandboxID), port)
	params := url.Values{}
	if useServerProxy != nil {
		params.Set("use_server_proxy", strconv.FormatBool(*useServerProxy))
	}
	if encoded := params.Encode(); encoded != "" {
		path += "?" + encoded
	}
	var resp Endpoint
	if err := c.doRequestCapture(ctx, "GET", path, nil, &resp, func(h http.Header) {
		resp.Origin = SandboxOrigin(h.Get(SandboxOriginHeader))
	}); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetSignedEndpoint retrieves a cryptographically signed endpoint URL for a
// sandbox port. The returned endpoint embeds an OSEP-0011 signed route token
// that expires at the given Unix epoch timestamp (seconds).
func (c *LifecycleClient) GetSignedEndpoint(ctx context.Context, sandboxID string, port int, expires int64) (*Endpoint, error) {
	path := fmt.Sprintf("/sandboxes/%s/endpoints/%d?expires=%d", url.PathEscape(sandboxID), port, expires)
	var resp Endpoint
	if err := c.doRequestCapture(ctx, "GET", path, nil, &resp, func(h http.Header) {
		resp.Origin = SandboxOrigin(h.Get(SandboxOriginHeader))
	}); err != nil {
		return nil, err
	}
	return &resp, nil
}
